fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/cursors/CursorLimitManagerTest.java

```diff
@@ -56,26 +61,42 @@ public void timeOut() {
 
     @Test
     public void testRecordScanLimiter() {
+        testRecordScanLimiterBase(false);
+        testRecordScanLimiterBase(true);
+    }
+
+    public void testRecordScanLimiterBase(boolean failOnLimitReached) {
         final int numberOfScans = 12;
         final RecordScanLimiter recordScanLimiter = new RecordScanLimiter(numberOfScans);
-        final CursorLimitManager manager = new CursorLimitManager(recordScanLimiter, false, null, null);
+        final CursorLimitManager manager = new CursorLimitManager(recordScanLimiter, failOnLimitReached, null, null);
 
         for (int i = 0; i < numberOfScans; i++) {
             assertTrue(manager.tryRecordScan());
             assertFalse(manager.isStopped());
             assertFalse(manager.getStoppedReason().isPresent());
         }
 
-        assertFalse(manager.tryRecordScan());
-        assertTrue(manager.isStopped());
-        assertEquals(RecordCursor.NoNextReason.SCAN_LIMIT_REACHED, manager.getStoppedReason().get());
+        if (failOnLimitReached) {
+            RecordCoreException ex = assertThrows(ScanLimitReachedException.class, manager::tryRecordScan);
+            assertThat(ex.getMessage(), containsString("SCAN_LIMIT_REACHED"));
```

Comment:
If you update the exception thrown to use logging properties to convey the reason the scan was stopped, you can use `TestHelpers.assertThrows()` to both assert the exception that is thrown as well as the log keys and values you expect.